primitives/api/proc-macro/src/impl_runtime_apis.rs

```diff
@@ -398,19 +401,63 @@ fn extend_with_runtime_decl_path(mut trait_: Path) -> Path {
 	trait_
 }
 
+// Dummy trait implementations are used to check if the versioned runtime implements all required
+// methods. `generate_versioned_api_traits` in `decl_runtime_apis.rs` generates a dummy trait for
+// each version containing all required methods that should be implemented. Here an implementation
+// for this dummy trait is generated by adding all methods defined in the api (and stripping their
+// implementations to save compiling time). If a method is missed the compiler will generate an
+// error which luckily is descriptive enough for the implementor.
+fn generate_dummy_trait_impl(initial: &ItemImpl, trait_api_ver: u64) -> ItemImpl {
+	let mut dummy_impl_ = initial.clone();
+	let mut dummy_trait = dummy_impl_
+		.trait_
+		.as_ref()
+		.expect("Api implementation should always contain a trait bound; qed")
+		.clone();
+	let mut trait_path_segment = dummy_trait
+		.1
+		.segments
+		.last()
+		.expect("Trait path should always contain at least one item; qed")
+		.clone();
+
+	trait_path_segment.ident = dummy_trait_name(&trait_path_segment.ident, trait_api_ver);
+
+	dummy_trait.1.segments.pop();
+	let pos = dummy_trait.1.segments.len();
+	dummy_trait.1.segments.insert(pos, trait_path_segment);
+
+	dummy_impl_.trait_ = Some(dummy_trait);
+
+	// attach dummy implementations to all methods
+	for i in &mut dummy_impl_.items {
+		match i {
+			ImplItem::Method(m) => attach_default_method_implementation(m),
+			_ => continue,
+		}
+	}
+
+	dummy_impl_
+}
+
 /// Generates the implementations of the apis for the runtime.
 fn generate_api_impl_for_runtime(impls: &[ItemImpl]) -> Result<TokenStream> {
 	let mut impls_prepared = Vec::new();
 
 	// We put `runtime` before each trait to get the trait that is intended for the runtime and
 	// we put the `RuntimeBlock` as first argument for the trait generics.
 	for impl_ in impls.iter() {
+		let trait_api_ver = extract_api_version(&impl_.attrs, impl_.span())?;
+
 		let mut impl_ = impl_.clone();
 		let trait_ = extract_impl_trait(&impl_, RequireQualifiedTraitPath::Yes)?.clone();
 		let trait_ = extend_with_runtime_decl_path(trait_);
 
 		impl_.trait_.as_mut().unwrap().1 = trait_;
 		impl_.attrs = filter_cfg_attrs(&impl_.attrs);
+		if let Some(api_ver) = trait_api_ver {
+			impls_prepared.push(generate_dummy_trait_impl(&impl_, api_ver));
+		}
```

Comment:
I don't get why you are doing this? `extend_with_runtime_decl_path` above should point to the correct trait that corresponds to the version as set by the user.

Reply:
Yes, now it reworked as you suggested.